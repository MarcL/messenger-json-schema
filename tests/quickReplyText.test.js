const ajv = require('./ajv');
const quickReplyText = require('../messenger/quickReplyText');

describe('Quick reply: text message', () => {
    let validate;
    
    beforeEach(() => {
        validate = ajv.compile(quickReplyText);
    });

    test('Valid message with string payload', () => {
        const givenMessage = {
            content_type: 'text',
            title: 'title',
            payload: 'payload'
        };

        validate(givenMessage);
        
        expect(validate.errors).toBeNull();
    });

    test('Valid message with number payload', () => {
        const givenMessage = {
            content_type: 'text',
            title: 'title',
            payload: 1001
        };

        validate(givenMessage);
        
        expect(validate.errors).toBeNull();
    });

    test('Valid message with image_url', () => {
        const givenMessage = {
            content_type: 'text',
            title: 'title',
            payload: 'payload',
            image_url: 'https://www.test.com/image.jpg'
        };

        validate(givenMessage);
        
        expect(validate.errors).toBeNull();
    });

    test('Message must contain image_url property if title is empty', () => {
        const givenMessage = {
            content_type: 'text',
            title: '',
            payload: 'payload',
            image_url: 'https://www.test.com/image.jpg'
        };

        validate(givenMessage);

        expect(validate.errors).toBeNull();
    });

    describe('Invalid message', () => {
        const invalidScenarios = [
            {
                testMessage: 'Message is not a string',
                givenMessage: {
                    content_type: 100,
                    title: 'title',
                    payload: 'payload'
                }
            },
            {
                testMessage: 'Message missing content_type property',
                givenMessage: {}
            },
            {
                testMessage: 'Message contains additional properties',
                givenMessage: {
                    content_type: 'text',
                    title: 'title',
                    payload: 'payload',
                    invalid: 'Invalid property'
                }
            },
            {
                testMessage: 'Message contains invalid content_type property',
                givenMessage: {
                    content_type: 'not_user_email',
                    title: 'title',
                    payload: 'payload'
                }
            },
            {
                testMessage: 'Message contains invalid payload property',
                givenMessage: {
                    content_type: 'text',
                    title: 'title',
                    payload: true
                }
            },
            {
                testMessage: 'Message contains invalid image_url property',
                givenMessage: {
                    content_type: 'text',
                    title: 'title',
                    payload: 'payload',
                    image_url: ''
                }
            }
        ];

        invalidScenarios.forEach(scenario => {
            const {testMessage, givenMessage} = scenario;

            test(testMessage, () => {        
                validate(givenMessage);

                expect(validate.errors).toMatchSnapshot();
            });
        });

        test('Message title is greater than 20 characters', () => {
            const title = new Array(21).fill('a').join('');
            const givenMessage = {
                content_type: 'text',
                title,
                payload: 'payload'
            };

            validate(givenMessage);

            expect(validate.errors).toMatchSnapshot();
        });

        test('Message payload is greater than 1000 characters', () => {
            const payload = new Array(1001).fill('a').join('');
            const givenMessage = {
                content_type: 'text',
                title: 'title',
                payload
            };

            validate(givenMessage);

            expect(validate.errors).toMatchSnapshot();
        });
    });
});
