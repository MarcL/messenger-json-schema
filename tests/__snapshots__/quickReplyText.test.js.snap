// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quick reply: text message Invalid message Message contains additional properties 1`] = `
Array [
  Object {
    "dataPath": "",
    "keyword": "additionalProperties",
    "message": "should NOT have additional properties",
    "params": Object {
      "additionalProperty": "invalid",
    },
    "schemaPath": "#/additionalProperties",
  },
]
`;

exports[`Quick reply: text message Invalid message Message contains invalid image_url property 1`] = `
Array [
  Object {
    "dataPath": "/image_url",
    "keyword": "format",
    "message": "should match format \\"url\\"",
    "params": Object {
      "format": "url",
    },
    "schemaPath": "#/properties/image_url/format",
  },
]
`;

exports[`Quick reply: text message Invalid message Message contains invalid payload type 1`] = `
Array [
  Object {
    "dataPath": "/payload",
    "keyword": "type",
    "message": "should be string",
    "params": Object {
      "type": "string",
    },
    "schemaPath": "#/properties/payload/anyOf/0/type",
  },
  Object {
    "dataPath": "/payload",
    "keyword": "type",
    "message": "should be number",
    "params": Object {
      "type": "number",
    },
    "schemaPath": "#/properties/payload/anyOf/1/type",
  },
  Object {
    "dataPath": "/payload",
    "keyword": "anyOf",
    "message": "should match some schema in anyOf",
    "params": Object {},
    "schemaPath": "#/properties/payload/anyOf",
  },
]
`;

exports[`Quick reply: text message Invalid message Message has a content_type which is invalid 1`] = `
Array [
  Object {
    "dataPath": "/content_type",
    "keyword": "enum",
    "message": "should be equal to one of the allowed values",
    "params": Object {
      "allowedValues": Array [
        "text",
      ],
    },
    "schemaPath": "#/properties/content_type/enum",
  },
]
`;

exports[`Quick reply: text message Invalid message Message has a content_type which is not a string 1`] = `
Array [
  Object {
    "dataPath": "/content_type",
    "keyword": "type",
    "message": "should be string",
    "params": Object {
      "type": "string",
    },
    "schemaPath": "#/properties/content_type/type",
  },
  Object {
    "dataPath": "/content_type",
    "keyword": "enum",
    "message": "should be equal to one of the allowed values",
    "params": Object {
      "allowedValues": Array [
        "text",
      ],
    },
    "schemaPath": "#/properties/content_type/enum",
  },
]
`;

exports[`Quick reply: text message Invalid message Message is missing a content_type property 1`] = `
Array [
  Object {
    "dataPath": "",
    "keyword": "required",
    "message": "should have required property 'content_type'",
    "params": Object {
      "missingProperty": "content_type",
    },
    "schemaPath": "#/required",
  },
]
`;

exports[`Quick reply: text message Invalid message Message payload is greater than 1000 characters 1`] = `
Array [
  Object {
    "dataPath": "/payload",
    "keyword": "maxLength",
    "message": "should NOT be longer than 1000 characters",
    "params": Object {
      "limit": 1000,
    },
    "schemaPath": "#/properties/payload/anyOf/0/maxLength",
  },
  Object {
    "dataPath": "/payload",
    "keyword": "type",
    "message": "should be number",
    "params": Object {
      "type": "number",
    },
    "schemaPath": "#/properties/payload/anyOf/1/type",
  },
  Object {
    "dataPath": "/payload",
    "keyword": "anyOf",
    "message": "should match some schema in anyOf",
    "params": Object {},
    "schemaPath": "#/properties/payload/anyOf",
  },
]
`;

exports[`Quick reply: text message Invalid message Message title is greater than 20 characters 1`] = `
Array [
  Object {
    "dataPath": "/title",
    "keyword": "maxLength",
    "message": "should NOT be longer than 20 characters",
    "params": Object {
      "limit": 20,
    },
    "schemaPath": "#/properties/title/maxLength",
  },
]
`;
